File: 366. Find Leaves of Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findLeaves(self, root: TreeNode) -> List[List[int]]:
        rev = {}
        stack = []
        node = root
        rev[root] = None
        leaves = []
        
        # create a reverse lookup
        # get leaf nodes
        while node or stack:
            if node:
                if node.left:
                    rev[node.left] = node
                if node.right:
                    rev[node.right] = node
                if not node.left and not node.right:
                    leaves.append(node)
                    
                stack.append(node.right)
                node = node.left
            else:
                node = stack.pop()
                
        # extract nodes
        sol = []
        while leaves:
            sub = []
            newleaves = []
            for leaf in leaves:
                sub.append(leaf.val)
                parent = rev[leaf]
                if parent:
                    if parent.left == leaf:
                        parent.left = None
                    else:
                        parent.right = None
                        
                    if not parent.left and not parent.right:
                        newleaves.append(parent)
                        
            leaves = newleaves
            sol.append(sub)
            
        return sol
        
                

File: 366. Find Leaves of Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 
 // O(n + n/2 + n/4 ..) = O(n) time complexity, O(n) space complexity
class Solution {
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> allLeaves = new ArrayList<List<Integer>>();
        if (root == null) {
            return allLeaves;
        }
        
        while (root.left != null || root.right != null) {
            allLeaves.add(extractLeaves(root));
        }
        
        allLeaves.add(new ArrayList<Integer>(Arrays.asList(root.val)));
        return allLeaves;
    }
    
    public List<Integer> extractLeaves(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        List<Integer> leaves = new ArrayList<Integer>(); 
        while (root != null || !stack.isEmpty()) {
            if (root != null) {
                if (root.left != null && root.left.left == null && root.left.right == null) {
                    leaves.add(root.left.val);
                    root.left = null;
                } else {
                    stack.add(root.left);
                }
                
                if (root.right != null && root.right.left == null && root.right.right == null) {
                    leaves.add(root.right.val);
                    root.right = null;
                } else {
                    stack.add(root.right);
                }
                
                root = null;
            } else {
                root = stack.pop();
            }
        }
        return leaves;
    }
}
